Detect systemd when inside a container.

judilsteve opened the pull request:
Modify IsSystemdService to function correctly
for containerised services.

Fix [#67203](https://github.com/dotnet/runtime/issues/67203)

Signed-off-by: james <judilsteve>

Commit:
Detect systemd when inside a container.

Modify IsSystemdService to function correctly
for containerised services.

Fix #67203

Signed-off-by: james <judilsteve>

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67288) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67288) before we can accept your contribution.<br/><br/>:x: judilsteve [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67288)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=67288) it.</sub>

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -13,6 +13,9 @@ namespace Microsoft.Extensions.Hosting.Systemd
     /// </summary>
     public static partial class SystemdHelpers
     {
+        internal const string NOTIFY_SOCKET_ENVVAR_KEY = "NOTIFY_SOCKET";
```
nit: drop the `_ENVVAR_KEY` suffix, and them it `private`.

judilsteve replied:
I made them internal so that the NOTIFY_SOCKET magic string could be shared with SystemdNotifier. Happy to revert that part and make the strings private if you'd prefer to just have the literal defined twice.

judilsteve replied:
Actioned

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -33,15 +36,32 @@ private static bool CheckParentIsSystemd()
             {
             }
 
+            // Convention for containers is to have the PID of the containerised process set to 1
```
Let's move this check before the existing one.

judilsteve replied:
Actioned

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -33,15 +36,32 @@ private static bool CheckParentIsSystemd()
+            // Convention for containers is to have the PID of the containerised process set to 1
+            if (Environment.ProcessId == 1)
+            {
+                if(!string.IsNullOrEmpty(Environment.GetEnvironmentVariable(NOTIFY_SOCKET_ENVVAR_KEY)))
```
We can combine these and write them as:
```
```
return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(NOTIFY_SOCKET_ENVVAR_KEY) ||
          !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(LISTEN_PID_ENVVAR_KEY);
```

judilsteve replied:
Actioned

judilsteve commented:
Would you prefer code review actions in a second commit, or to have the original commit re-made and force pushed?

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -33,15 +36,32 @@ private static bool CheckParentIsSystemd()
                 string ppidString = parentPid.ToString(NumberFormatInfo.InvariantInfo);
                 byte[] comm = File.ReadAllBytes("/proc/" + ppidString + "/comm");
-                return comm.AsSpan().SequenceEqual(Encoding.ASCII.GetBytes("systemd\n"));
+                if (comm.AsSpan().SequenceEqual(Encoding.ASCII.GetBytes("systemd\n")))
```
After moving the PID 1 check before this, this can go back to what it was.

judilsteve replied:
Actioned

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdNotifier.cs:
```diff
@@ -48,7 +46,7 @@ public void Notify(ServiceState state)
         private static string GetNotifySocketPath()
         {
-            string socketPath = Environment.GetEnvironmentVariable(NOTIFY_SOCKET);
+            string socketPath = Environment.GetEnvironmentVariable(SystemdHelpers.NOTIFY_SOCKET_ENVVAR_KEY);
```
I'd keep the `private const string` in this class.

judilsteve replied:
Actioned

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -33,15 +36,32 @@ private static bool CheckParentIsSystemd()
             {
             }
 
+            // Convention for containers is to have the PID of the containerised process set to 1
```
suggestion for the comment:
> To support containerized systemd services, check if we're the main process (PID 1) and if there are systemd environment variables defined for notifying the service manager, or passing listen handles.

judilsteve replied:
Actioned

tmds commented:
> Would you prefer code review actions in a second commit, or to have the original commit re-made and force pushed?

Contributors work according to their own preferences. My personal preference is to have separate commits so I can keep track of changes more easily.

Commit:
Code review actions.

Implement review actions for pull request 67288.

Signed-off-by: james <judilsteve>

Commit:
Fix trailing whitespace.

Self-review action for pull request 67288.

Signed-off-by: james <judilsteve>

Commit:
Fix build errors.

Fix incorrect variable names.

Signed-off-by: james <judilsteve>

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -30,9 +33,19 @@ private static bool CheckParentIsSystemd()
             try
             {
                 // Check whether our direct parent is 'systemd'.
+                // This will not work for containerised applications, since our direct parent will be some container runtime
```
I think we can remove this comment.

Commit:
Code review actions.

Remove superfluous comment.

Signed-off-by: james <judilsteve>

tmds commented:
@judilsteve this is looking good!

Can you verify on your local machine that this works as expected?

You should be able to use a self-contained app that you build using https://dotnet.microsoft.com/en-us/download/dotnet/7.0, and overwrite the systemd hosting dll with the one you've built.

judilsteve commented:
I'll have to leave testing the DLL for another day; it's getting late where I am.

tmds commented:
No problem, you can test when you find some time for it.

jkotas commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -13,6 +13,9 @@ namespace Microsoft.Extensions.Hosting.Systemd
     /// </summary>
     public static partial class SystemdHelpers
     {
+        private const string NOTIFY_SOCKET = "NOTIFY_SOCKET";
```
Nit: This does not follow the repo naming conventions. https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md We use PascalCasing to name all our constant local variables and fields.

judilsteve replied:
Normally I would have just used a literal, but I was following the lead of [SystemNotifier](https://source.dot.net/#Microsoft.Extensions.Hosting.Systemd/SystemdNotifier.cs). Unlucky that I chose a bad example, I guess.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -30,6 +33,15 @@ private static bool CheckParentIsSystemd()
+            // To support containerized systemd services, check if we're the main process (PID 1)
+            // and if there are systemd environment variables defined for notifying the service
+            // manager, or passing listen handles.
+            if (Environment.ProcessId == 1)
+            {
+                return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(NOTIFY_SOCKET)) ||
+                    !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(LISTEN_PID));
+            }
+
```
(nit) The name of this method is `CheckParentIsSystemd()`, but this is doing more than that. Can we update this method name to be something more like `GetIsSystemdService()`?

judilsteve replied:
Good spot, next commit will fix.

Commit:
Code review actions.

Use literal instead of constant.
Update function name to reflect new behaviour.

Signed-off-by: james <judilsteve>

judilsteve commented:
I've updated my test case to use dotnet 7 and inject the patched DLL (see [here](https://github.com/judilsteve/FdListenTest/commit/62f486b5cbf01ec1f94ad05c1c24f83480721dc7)). `UseSystemd()` now functions correctly when running inside the container:

```
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: ISystemdNotifier is not enabled
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: SystemdLifetime is not enabled
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: Calling nameof(SystemHostBuilderExtensions.UseSystemd)
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: ISystemdNotifier is enabled
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: SystemdLifetime is enabled
Mar 30 18:08:30 lappy-486 fd-listen-test[21088]: <6>Microsoft.Hosting.Lifetime[14] Now listening on: http://[::]:4242
```

Systemd still doesn't get the "service ready" message, but that seems like an issue with either podman or systemd, not dotnet: Even jumping into a shell inside the container and `socat`ing `READY=1` into the notify socket doesn't work.

judilsteve commented:
Have also tested the non-containerised version in 7.0 with the patched DLL: It still works as before.

judilsteve commented:
A few of the CI checks are failling with the following error:

![image](https://user-images.githubusercontent.com/4328790/160814880-92199498-c1d9-41a8-97a3-bbe8ec008faa.png)

As far as I can tell, the property should exist:
https://docs.microsoft.com/en-us/dotnet/api/system.environment.processid?view=net-7.0

Could someone please provide guidance on how to fix this?

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -20,16 +20,25 @@ public static partial class SystemdHelpers
         /// <returns><c>True</c> if the current process is hosted as a systemd Service, otherwise <c>false</c>.</returns>
         public static bool IsSystemdService()
-            => _isSystemdService ?? (bool)(_isSystemdService = CheckParentIsSystemd());
+            => _isSystemdService ?? (bool)(_isSystemdService = GetIsSystemdService());
```
While we are here changing this line, this can be

```suggestion
            => _isSystemdService ??= GetIsSystemdService();
```

Commit:
Fix build on .NET <5.0.

Adds conditional compilation to fall back
to Process.GetCurrentProcess().Id when
Environment.ProcessId is not available.

Also implements a minor code-review suggestion.

Signed-off-by: james <judilsteve>

Commit:
Add conditional using statement.

Should fix build on .NET <5

Signed-off-by: james <judilsteve>

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -20,16 +23,31 @@ public static partial class SystemdHelpers
+            if (processId == 1)
+            {
+                return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("NOTIFY_SOCKET")) ||
+                    !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("LISTEN_PID"));
```
nit: align the `!`.

tmds commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -20,16 +23,31 @@ public static partial class SystemdHelpers
+            int processId = Process.GetCurrentProcess().Id;
+#endif
+
+            if (processId == 1)
```
@eerhardt @judilsteve 

As I mentioned in the corresponding issue, this is a good way to check the current process is main process in the container.

In case we're not pid 1, we could extend it with checking if the pid 1 environment (`/proc/1/environ`) contains a `container` envvar. This file may be inaccessible. That is the way systemd itself checks if it runs as systemd in a container. And systemd sets this for its containerized services.

If we want to be looser, we could drop the checks and just rely on the envvars.

I'm fine with the implementation as is. I want to mention the options.

judilsteve commented:
Looks like something is up with the [iOS build pipeline](https://dev.azure.com/dnceng/public/_build/results?buildId=1692344&view=logs&jobId=a5078f86-b345-5a4a-85ee-f64916152c6f&j=a5078f86-b345-5a4a-85ee-f64916152c6f&t=b1d0531b-2b6c-5e64-cc59-e2e1ffcc72bf). Certificate error when attempting to report status to something called Helix? This doesn't seem like it would be related to this PR.

Commit:
Align multi-line OR statement.

Code review action.

Signed-off-by: james <judilsteve>

judilsteve commented:
Build is green and I think all review items have been actioned.

eerhardt reviewed (approved):
I think this change looks fine. @tmds - if you are happy with it, I can merge it.

tmds commented:
Thanks @judilsteve!

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -5,6 +5,9 @@
 using System.Globalization;
 using System.IO;
 using System.Text;
+#if !NET5_0_OR_GREATER
```
```suggestion
#if !NETCOREAPP
```

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHelpers.cs:
```diff
@@ -20,16 +23,31 @@ public static partial class SystemdHelpers
+            // To support containerized systemd services, check if we're the main process (PID 1)
+            // and if there are systemd environment variables defined for notifying the service
+            // manager, or passing listen handles.
+#if NET5_0_OR_GREATER
```
```suggestion
#if NETCOREAPP
```

eerhardt reviewed (commented):
On .NETCoreApp, this library only supports 6.0 and 7.0.  So there is no need for the more complicated `NET5_0_OR_GREATER` define. See Commit 1 of https://github.com/dotnet/runtime/pull/65847.

eerhardt committed:
Apply suggestions from code review

The minimum supported .NETCoreApp version in the repository is 6.0, so just using NETCOREAPP.

eerhardt commented:
Failures are unrelated to this change. They look like https://github.com/dotnet/runtime/issues/67159.

Merging.
